File: shared/protocol/teleop_protocol/enums.py
from enum import Enum


class Role(str, Enum):
    """Participant roles within a teleop session."""

    OPERATOR = "operator"
    LEADER = "leader"
    FOLLOWER = "follower"


class SessionState(str, Enum):
    """Session/teleop state machine states.

    Transitions (see docs/protocol/messages.md for the full table):
      DISCONNECTED -> CONNECTED -> READY -> CLAIMED -> ARMED
      ARMED -> CLUTCHED | FROZEN
      ANY -> ESTOPPED
      ESTOPPED -> READY (only via explicit reset)
    """

    DISCONNECTED = "DISCONNECTED"
    CONNECTED = "CONNECTED"
    READY = "READY"
    CLAIMED = "CLAIMED"
    ARMED = "ARMED"
    CLUTCHED = "CLUTCHED"
    FROZEN = "FROZEN"
    ESTOPPED = "ESTOPPED"


class RobotMode(str, Enum):
    """Reported follower robot mode (independent of session ownership state)."""

    IDLE = "IDLE"
    TRACKING = "TRACKING"
    HOMING = "HOMING"
    FROZEN = "FROZEN"
    ESTOPPED = "ESTOPPED"
    ERROR = "ERROR"


class PeerMode(str, Enum):
    """WebRTC connectivity mode as reported by ICE candidate pair selection."""

    UNKNOWN = "UNKNOWN"
    P2P = "P2P"
    RELAY = "RELAY"


class GripperCommand(str, Enum):
    OPEN = "open"
    CLOSE = "close"


class ErrorCode(str, Enum):
    PROTOCOL_VERSION_MISMATCH = "protocol_version_mismatch"
    NOT_OWNER = "not_owner"
    NOT_ARMED = "not_armed"
    ESTOPPED = "estopped"
    STALE_PACKET = "stale_packet"
    SEQUENCE_REPLAY = "sequence_replay"
    INVALID_MESSAGE = "invalid_message"
    SESSION_NOT_FOUND = "session_not_found"
    SESSION_FULL = "session_full"
    UNAUTHORIZED = "unauthorized"
    INTERNAL_ERROR = "internal_error"

File: shared/types/src/enums.ts
/** Protocol version. Bump in lockstep with shared/protocol/teleop_protocol/version.py */
export const PROTOCOL_VERSION = 1;

export enum Role {
  Operator = "operator",
  Leader = "leader",
  Follower = "follower",
}

/**
 * Session/teleop state machine states.
 *
 * Transitions:
 *   DISCONNECTED -> CONNECTED -> READY -> CLAIMED -> ARMED
 *   ARMED -> CLUTCHED | FROZEN
 *   ANY -> ESTOPPED
 *   ESTOPPED -> READY (only via explicit reset)
 */
export enum SessionState {
  Disconnected = "DISCONNECTED",
  Connected = "CONNECTED",
  Ready = "READY",
  Claimed = "CLAIMED",
  Armed = "ARMED",
  Clutched = "CLUTCHED",
  Frozen = "FROZEN",
  Estopped = "ESTOPPED",
}

export enum RobotMode {
  Idle = "IDLE",
  Tracking = "TRACKING",
  Homing = "HOMING",
  Frozen = "FROZEN",
  Estopped = "ESTOPPED",
  Error = "ERROR",
}

export enum PeerMode {
  Unknown = "UNKNOWN",
  P2P = "P2P",
  Relay = "RELAY",
}

export enum GripperCommand {
  Open = "open",
  Close = "close",
}

export enum ErrorCode {
  ProtocolVersionMismatch = "protocol_version_mismatch",
  NotOwner = "not_owner",
  NotArmed = "not_armed",
  Estopped = "estopped",
  StalePacket = "stale_packet",
  SequenceReplay = "sequence_replay",
  InvalidMessage = "invalid_message",
  SessionNotFound = "session_not_found",
  SessionFull = "session_full",
  Unauthorized = "unauthorized",
  InternalError = "internal_error",
}
